src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.InProcOrRemoteHostAnalyzerRunner.cs

```diff
@@ -12,145 +12,219 @@
+            bool forceExecuteAllAnalyzers,
+            bool logPerformanceInfo,
+            bool getTelemetryInfo,
+            CancellationToken cancellationToken)
+        {
+            var result = await AnalyzeCoreAsync().ConfigureAwait(false);
+            Debug.Assert(getTelemetryInfo || result.TelemetryInfo.IsEmpty);
+            return result;
 
-            public async Task<DiagnosticAnalysisResultMap<DiagnosticAnalyzer, DiagnosticAnalysisResult>> AnalyzeAsync(CompilationWithAnalyzers compilation, Project project, bool forcedAnalysis, CancellationToken cancellationToken)
+            async Task<DiagnosticAnalysisResultMap<DiagnosticAnalyzer, DiagnosticAnalysisResult>> AnalyzeCoreAsync()
             {
-                Contract.ThrowIfFalse(compilation.Analyzers.Length != 0);
+                Contract.ThrowIfFalse(!compilationWithAnalyzers.Analyzers.IsEmpty);
 
                 var workspace = project.Solution.Workspace;
-                var service = workspace.Services.GetService<IRemoteHostClientProvider>();
-                if (service == null)
+                if (workspace.Services.GetService<IRemoteHostClientProvider>() is { } service &&
+                    await service.TryGetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false) is { } remoteHostClient)
                 {
-                    // host doesn't support RemoteHostService such as under unit test
-                    return await AnalyzeInProcAsync(compilation, project, client: null, cancellationToken).ConfigureAwait(false);
+                    return await AnalyzeOutOfProcAsync(documentAnalysisScope, project, compilationWithAnalyzers, remoteHostClient,
+                        forceExecuteAllAnalyzers, logPerformanceInfo, getTelemetryInfo, cancellationToken).ConfigureAwait(false);
                 }
 
-                var remoteHostClient = await service.TryGetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false);
-                if (remoteHostClient == null)
-                {
-                    // remote host is not running. this can happen if remote host is disabled.
-                    return await AnalyzeInProcAsync(compilation, project, client: null, cancellationToken).ConfigureAwait(false);
-                }
-
-                // out of proc analysis will use 2 source of analyzers. one is AnalyzerReference from project (nuget). and the other is host analyzers (vsix) 
-                // that are not part of roslyn solution. these host analyzers must be sync to OOP before hand by the Host. 
-                return await AnalyzeOutOfProcAsync(remoteHostClient, compilation, project, forcedAnalysis, cancellationToken).ConfigureAwait(false);
+                // We need to execute InProc due to one of the following reasons:
+                //  1. The host doesn't support RemoteHostService (such as under unit test) OR
+                //  2. Remote host is not running (this can happen if remote host is disabled).
```

Comment:
>2. Remote host is not running (this can happen if remote host is disabled) [](start = 20, length = 74)

[2] can't happen. We don't allow disabling remote host anymore.

Reply:
I removed the comment entirely as we are now directly using `RemoteHostClient.TryGetClientAsync`